Refactor documentation

xkubov opened the pull request:
This MR contains large refactor of IS Core documentation providing detailed overview, architecture and thorough examples on how to use the library.

Commit:
docs: initialize documentation improvements and README updates

- Add documentation build instructions to README
- Update Sphinx configuration

Commit:
docs: add comprehensive guides for architecture, forms, menu, REST, and views

- Add architecture overview and request flow documentation
- Add forms guide with fieldsets and validation
- Add menu generation documentation
- Add REST resources documentation
- Add views documentation with table/detail/form views

Commit:
docs: fix typos and formatting across multiple files

- Fix typos in architecture, filters, index, installation, permissions

Commit:
docs(views): add views overview comparison table

- Add table comparing different view types

Commit:
docs(architecture): expand cores overview section

- Add detailed core types explanation

Commit:
docs(permissions): add quick reference section

- Add permissions quick reference guide

Commit:
docs: add cross-references between related sections

- Link architecture, cores, forms, menu, permissions, rests, views

Commit:
docs(filters,utils): expand filters and utils documentation

- Add 459 lines of filter examples and utilities

Commit:
refactor(docs): rename list_display to list_fields throughout

- Update terminology across architecture, cores, filters, utils, views

Commit:
refactor(docs): reorganize home core information

- Move home core documentation from architecture to cores section

Commit:
docs(cores): address documentation TODOs

- Clean up and improve cores documentation

Commit:
docs: remove obsolete DynamoDB documentation

- Delete dynamodb.rst and update index

Commit:
docs: configure API reference generation

- Add autodoc configuration to Sphinx

Commit:
docs: remove obsolete DynamoDB documentation

- Delete dynamodb.rst and update index

Commit:
docs: configure API reference generation

- Add autodoc configuration to Sphinx
- Update index with reference section

Commit:
build(docs): ensure documentation is buildable

- Fix Sphinx configuration issues
- Update example requirements
- Fix cores.rst imports

Commit:
docs: add production examples to cores, filters, REST, and views

- Add 1135 lines of real-world examples from example_customers

Commit:
docs: add additional examples to cores, filters, REST, and views

Commit:
docs: add architecture diagrams

- Add core-registration.png and iscore-arch.png

xkubov commented on README.md:
```diff
@@ -33,8 +33,34 @@ Docs
 
-Contribution
-------------
+Documentation
```
We should add a section on how to install virtulenv before this. This section depends on activating virtualenv from the `example` app

xkubov commented on README.md:
```diff
@@ -33,8 +33,34 @@ Docs
 To run Livereload for Sphinx you need [livereload](https://pypi.python.org/pypi/livereload) `pip install livereload`.
-After installing simply call `make htmllivereload` and open [http://localhost:5500/](http://localhost:5500/).
\ No newline at end of file
+After installing simply call `make htmllivereload` and open [http://localhost:5500/](http://localhost:5500/).
```
this step doesn't work - we should either make it work or discard this section

xkubov commented on docs/architecture.rst:
```diff
@@ -0,0 +1,191 @@
+**Resources (REST)**
+  Classes that define REST endpoints (GET/POST/PUT/DELETE) acting as the backend for UI actions. Resources provide the data and actions that views consume.
+
+Request Flow
+------------
+
+A typical request in Django IS Core follows this pattern:
+
+.. image:: .images/iscore-arch.png
+   :alt: IS Core Architecture Diagram
+   :align: center
+   :width: 200%
+
+The flow in detail:
+
+1. **Initial Request**: Browser hits a URL → View returns HTML template with links to JS/static files
+2. **Static Assets**: Browser loads JavaScript, CSS, and images from static files
+3. **Data Loading**: JavaScript calls the appropriate Core's REST endpoint for data (e.g., to populate a table)
+4. **Interactive Updates**: Further interactions (filtering, row actions) go through REST → JS updates the DOM
```
This should be updated so that it is clear that this is not typical request but example request that showcases  server-side and client-side rendering flow.

xkubov commented on docs/architecture.rst:
```diff
@@ -0,0 +1,191 @@
+Types of Core
+-------------
+
+Django IS Core provides several types of Core classes that can be combined based on your needs:
```
Add a comment that cores are described in cores section. Where cores are described in more detail

xkubov commented on docs/architecture.rst:
```diff
@@ -0,0 +1,191 @@
+| **Use Case**              | **Core Type**             | **What You Get**          |
++===========================+===========================+===========================+
+| Full admin interface      | DjangoUiRestCore          | UI views + REST API       |
++---------------------------+---------------------------+---------------------------+
+| REST API only             | DjangoRestCore            | API endpoints only        |
++---------------------------+---------------------------+---------------------------+
+| UI with external API      | DjangoUiCore              | UI views (REST elsewhere) |
++---------------------------+---------------------------+---------------------------+
+| Custom frontend (React)   | DjangoRestCore            | API for your frontend     |
++---------------------------+---------------------------+---------------------------+
+| Static pages only         | UiCore                    | Simple server-side views  |
++---------------------------+---------------------------+---------------------------+
+
+UI Core
+^^^^^^^
+
+Handles server-side rendered views without REST queries during interaction. Useful for simple, static pages.
+
+REST Core
+^^^^^^^^^
+
+Generates REST API over models providing CRUD operations, pagination, filtering, etc. Can be used independently to build custom frontends (e.g., React, Vue).
+
+Django UI Core
+^^^^^^^^^^^^^^
+
+A UI Core specialized for Django models providing default list/detail/create views automatically.
+
+Django REST Core
+^^^^^^^^^^^^^^^^
+
+Default REST API over Django models with automatic CRUD endpoints.
+
+Combined Core (DjangoUiRestCore)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The most common case - combines UI + REST in a single Core. Provides both the admin interface and the REST API that powers it.
+
+.. tip::
+   **Recommended**: Use ``DjangoUiRestCore`` for most use cases. It gives you both admin UI and REST API with a single class definition.
```
can we add links to the references api?

xkubov commented on docs/architecture.rst:
```diff
@@ -0,0 +1,191 @@
+    class ArticleCore(DjangoUiRestCore):
+        model = Article
+        list_fields = ('title', 'author', 'created_at')
+        form_fields = ('title', 'content', 'author')
```
`form_fields` should be ideally just `fields`

xkubov commented on docs/architecture.rst:
```diff
@@ -0,0 +1,191 @@
+Registration and Auto-discovery
+--------------------------------
+
+Django IS Core uses automatic registration through metaclasses and singletons:
```
Specify, that the only thing that library users need to do is create `cores.py` in their application registred in `INSTALLED_APPS`.

xkubov commented on docs/architecture.rst:
```diff
@@ -0,0 +1,191 @@
+Generic Resource
+^^^^^^^^^^^^^^^^
+
+A custom endpoint where you implement get/post/put/delete methods and register it via ``get_rest_patterns`` on a specific path.
```
can we also describe where this resource needs to be registred? in core? Or even better add link to the section that describes it in more detail.

xkubov commented on docs/cores.rst:
```diff
@@ -713,27 +848,78 @@ Options
+        @filter_class(AppliedPromocodeIDFilter)
+        @property
+        def applied_promocodes_id(self):
+            return None  # Return value not used, filter handles the logic
```
not returning any value means `rest_extra_fields` will always return `None` for this property. We should specify that.

xkubov commented on docs/cores.rst:
```diff
@@ -713,27 +848,78 @@ Options
+
+- **Performance**: Expensive computed fields are only calculated when needed
+- **API flexibility**: Frontend can request additional data without changing default payload
+- **Separation**: REST-only data doesn't clutter UI field definitions
```
Remove this section.

xkubov commented on docs/cores.rst:
```diff
@@ -774,3 +955,281 @@ The atribute inside response has named ``_class_names``.
+
+
+    class CustomerCommentRestPattern(CustomerCommentBasePattern, RestPattern):
+        pass
```
remove this section

xkubov commented on docs/cores.rst:
```diff
@@ -774,3 +955,281 @@ The atribute inside response has named ``_class_names``.
+- ``CustomerCommentCore`` in ``cores/customer/comment.py`` handles comments
+
+Each gets full UI, REST API, export, and independent evolution—all while maintaining the nested URL structure through custom patterns.
+
```
remove this section

xkubov commented on docs/filters.rst:
```diff
@@ -1,35 +1,593 @@
+        def word_count(self, obj):
+            return len(obj.content.split())
+        word_count.filter = True  # Enable filtering
+        word_count.filter_by = 'content'  # Filter by content field
```
Remove this section

xkubov commented on docs/filters.rst:
```diff
@@ -1,35 +1,593 @@
+                PaymentMethodFilter,
+                enum_class=PaymentMethodEnum
+            )
+            return manager
```
remove this section.

xkubov commented on docs/forms.rst:
```diff
@@ -1,3 +1,674 @@
+- Thumbnail preview of current image
+- "Clear" checkbox to remove existing file
+- File size display
+- Upload progress indication
```
Remove this section.

xkubov commented on docs/forms.rst:
```diff
@@ -1,3 +1,674 @@
+
+6. **Use custom widgets** for specialized input types (date pickers, file uploads, etc.)
+
+7. **Test form validation** thoroughly, especially for complex multi-field validation
```
Remove this section.

xkubov commented on docs/forms.rst:
```diff
@@ -1,3 +1,674 @@
+
+            return tuple(fieldsets)
+
+Using fieldsets_postfix
+------------------------
+
+Append fieldsets to inherited configurations:
+
+.. code-block:: python
+
+    from user_comments.contrib.is_core.cores import CommentCoreMixin
+
+
+    class CustomerDetailView(CommentCoreMixin, DjangoDetailFormView):
+        # Base fields from parent
+        fields = ('first_name', 'last_name', 'email')
+
+        # Append after parent's fieldsets
+        fieldsets_postfix = CommentCoreMixin.notes_fieldset + (
+            (_l('Bank Accounts'), {
+                'inline_view': CustomerBankAccountInlineTableView,
+            }),
+        )
+
+This preserves the parent's fieldset configuration while adding your own.
```
Remove this section.

xkubov commented on docs/menu.rst:
```diff
@@ -1,12 +1,605 @@
+                ))
+                menu.append(settings)
+
+            return menu
```
Remove this section

xkubov commented on docs/menu.rst:
```diff
@@ -1,12 +1,605 @@
+                    icon='settings'
+                ))
+
+            return menu
```
Remove this section

xkubov commented on docs/menu.rst:
```diff
@@ -1,12 +1,605 @@
+
+9. **Consider mobile**: Ensure menu works well on small screens
+
+10. **Document custom menus**: If using a custom generator, document the structure for other developers
```
Remove this section.

xkubov commented on docs/rests.rst:
```diff
@@ -1,7 +1,996 @@
+REST Resources
+==============
+
+Django IS Core provides a powerful REST API system built on top of django-pyston. Resources define REST endpoints that serve as the backend for UI actions and can be used independently to build custom frontends.
```
Tone down the marketing. Be more pragmatic, narrate what it is, don't try to sell how "powerful" it is.

xkubov commented on docs/rests.rst:
```diff
@@ -1,7 +1,996 @@
+
+9. **Version your API** if you plan to make breaking changes
+
+10. **Test REST endpoints** thoroughly, including error cases and edge conditions
```
Remove this section.

Commit:
docs: address PR review feedback

- Remove marketing language and best practices sections
- Add virtualenv installation to README
- Clarify examples and add API references
- Remove 521 lines of redundant/marketing content

xkubov commented on docs/architecture.rst:
```diff
@@ -0,0 +1,201 @@
+
+An internal "model resource" providing ready-made CRUD over a model. Behavior can be extended or overridden.
+
+List Actions
+^^^^^^^^^^^^
+
+The backend returns a list of actions for each row via ``get_list_actions``:
+
+**UI actions**
+  e.g., redirect to detail view (without calling backend)
+
+**API actions**
+  e.g., delete action that calls REST; JavaScript executes the call and handles UI consequences (hide row, refresh, etc.)
+
+Example::
+
+    class ArticleCore(UIRestCore):
+        model = Article
+
+        def get_list_actions(self, obj, request):
+            actions = []
+            if request.user.has_perm('articles.change_article'):
+                actions.append({
+                    'name': 'edit',
+                    'url': self.get_edit_url(obj),
+                    'type': 'ui'
+                })
+            if request.user.has_perm('articles.delete_article'):
+                actions.append({
+                    'name': 'delete',
+                    'url': self.get_api_url('delete', obj.pk),
+                    'type': 'api'
+                })
+            return actions
```
add references to the place where it is more thoroughly described.

xkubov commented:
<img width="1892" height="971" alt="image" src="https://github.com/user-attachments/assets/3c628dd2-706b-4650-a419-722b2b9f8458" />

vichi99 commented on README.md:
````diff
@@ -33,8 +33,88 @@ Docs
+3. Build the HTML documentation:
+   ```bash
+   cd docs
+   make html
````
It doesnt work.

```
uv venv --python 3.11 
uv init
uv add sphinx sphinx_rtd_theme
source .venv/bin/activate
cd docs
make html
```

ModuleNotFoundError: No module named 'django'

I had to install 'django'. Could you add it to the README, including the version?


Then I was able to generate the docs, it works on `python -m http.server 8000 --directory .build/html`, ok.
However, I got these warnings. Could you please fix or describe this in the README? 

<img width="827" height="476" alt="Image" src="https://github.com/user-attachments/assets/a247da23-12b4-4a2f-b61e-a44cde3741d9" />

xkubov replied:
Hm, I admit the instructions are not clear. I'll clear that up. 

btw. you need to use venv from the `example` app. By using that you'll have the required django dep.

xkubov replied:
It should be clear how to setup the project now. Can you try?

vichi99 replied:
It is clearer now and it works. However, there are still warnings when building the documentation. Should we leave them like this?

<img width="999" height="456" alt="image" src="https://github.com/user-attachments/assets/b0d0c3d0-e1c9-413e-a540-13f68f579525" />

xkubov replied:
I fixed those warnings.

vichi99 commented on README.md:
````diff
@@ -33,8 +33,88 @@ Docs
+
+2. Install and set up the application:
+   ```bash
+   make install
````
After `make install` I got.

```
find . -name "*.pyc" -delete;
find . -type d -empty -delete;
rm -rf var/ve
rm -rf /Users/jan.vicha/Projects/work/skippay/django-is-core/example/var
virtualenv -p python3 var/ve
bash: virtualenv: command not found
make: *** [initvirtualenv] Error 127
```

xkubov replied:
I'll take a look at those instructions

vichi99 replied:
It is working now.

vichi99 commented on docs/views.rst:
```diff
@@ -1,60 +1,848 @@
+Allows bulk editing of multiple objects at once.
+
+**Use case:** Change status of multiple articles from "draft" to "published"
+
```
- There is no example to use BulkChangeFormView. However, we didnt use it in Core, so I guess its fine.
- It is not mentioned that bulk_change_enabled is set to True.
- bulk_change_fields is missing.

xkubov replied:
added

vichi99 replied:
checked

vichi99 commented on docs/menu.rst:
- Missing a description for CollapsibleMenuItem and an example. We use this as a BackendGroupMenuItem in Core.
- nitpick: Missing a description for MenuGenerator and an example. There is only for CustomMenuGenerator, but maybe that is purpose.

xkubov replied:
I think it is not required. `CollapsibleMenuItem` is redundant class. It just initializes MenuItem with `collapsible=True`.  I would much rather not have it in is core at all.

vichi99 replied:
Okay, we can discuss this in another task.

vichi99 commented on docs/cores.rst:
It would be good to add an explanation for Mixin Composition.
- The order of mixins (Method Resolution Order).
- How mixins work together.
- Best practices.

xkubov replied:
We can in separate MR. I think mixin part is nice to have. Purpose of this MR was primarily to add instructions on how to run the project and add architecture description.

vichi99 replied:
Okay, we can discuss this in another task.

vichi99 reviewed (commented):
Good job overall, just a few minor things to fix.

Commit:
docs: add API reference documentation and fix autodoc errors

Add comprehensive API reference documentation for Core, View, and Resource classes. Fix all Sphinx autodoc build errors by correcting class references, module paths, and handling missing dependencies.

Changes:
- Add docs/reference/ section with cores, views, and resources references
- Remove non-existent class references (BaseModelResource, DjangoModelAddFormView, etc.)
- Fix inline view module paths (form_views → inline_form_views, table_views → inline_table_views)
- Comment out CeleryDjangoCoreResource (requires security library)
- Replace serializers.Serializer with note about Pyston framework
- Update base view classes to use existing DefaultCoreViewMixin and DefaultModelCoreViewMixin

Commit:
docs: fix Sphinx build warnings

Fix remaining documentation build warnings:
- Remove non-existent .static directory from html_static_path
- Fix malformed inline literal in permissions.rst (add space before ::)
- Add custom_url_patterns label to rests.rst for Generic Resource section
- list_actions reference already exists in cores.rst (warning should be resolved)

vichi99 commented on docs/reference/resources.rst:
```diff
@@ -0,0 +1,47 @@
+   :members:
+   :show-inheritance:
+
+.. autoclass:: is_core.rest.resource.DjangoResource
```
References to is_core.rest.resource.BaseResource and is_core.rest.resource.DjangoResource are incorrect - these classes are imported from pyston.resource, not defined in is_core.rest.resource.

xkubov replied:
fixed.

Commit:
fix: Refernces + project home
